xla/service/gpu/openxla/compiler.cc

```diff
@@ -0,0 +1,235 @@
+                                 iree_compiler_invocation_t* inv)
+    : session_(session), inv_(inv) {}
+
+OpenXlaCompiler::~OpenXlaCompiler() {
+  if (error_) {
+    ireeCompilerErrorDestroy(error_);
+  }
+
+  ireeCompilerInvocationDestroy(inv_);
+  ireeCompilerSessionDestroy(session_);
+
+  if (output_) {
+    ireeCompilerOutputDestroy(output_);
+  }
+}
+
+bool OpenXlaCompiler::ParseSourceBuffer(std::string_view buffer) {
+  iree_compiler_source_t* source;
+  auto* error = ireeCompilerSourceWrapBuffer(
+      session_, "<jit>", buffer.data(), buffer.size(),
+      /*isNullTerminated=*/false, &source);
+  if (error) {
+    SetError(error);
+    return false;
+  }
+
+  return ireeCompilerInvocationParseSource(inv_, source);
+}
+
+bool OpenXlaCompiler::SetFlag(const char* flag) {
+  auto* error = ireeCompilerSessionSetFlags(session_, 1, &flag);
+  if (error) {
+    SetError(error);
+    return false;
+  }
+  return true;
+}
+
+std::unique_ptr<OpenXlaCompiler::Bytecode>
+OpenXlaCompiler::CompileStandardPipeline() {
+  if (!ireeCompilerInvocationPipeline(inv_, IREE_COMPILER_PIPELINE_STD)) {
+    return nullptr;
+  }
+
+  iree_compiler_error_t* error = ireeCompilerOutputOpenMembuffer(&output_);
+  if (error) {
+    SetError(error);
+    return nullptr;
+  }
+
+  error = ireeCompilerInvocationOutputVMBytecode(inv_, output_);
+  if (error) {
+    SetError(error);
+    return nullptr;
+  }
+
+  void* output_data = nullptr;
+  uint64_t size;
+  error = ireeCompilerOutputMapMemory(output_, &output_data, &size);
+  if (error) {
+    SetError(error);
+    return nullptr;
+  }
+
+  // Transfer the output_ to Bytecode since the mapping is only
+  // valid for the life of the output.
+  iree_compiler_output_t* local_output = output_;
+  output_ = nullptr;
+  return std::make_unique<Bytecode>(local_output, output_data, size);
+}
+
+//===-----------------------------------------------------------------------===/
+// Loading OpenXlaCompiler from a library
+//===-----------------------------------------------------------------------===/
+
+static bool InitializeCompilerForProcess(const std::string& library_path) {
+  if (!ireeCompilerLoadLibrary(library_path.c_str())) {
+    return false;
+  }
+
+  ireeCompilerGlobalInitialize();
+  return true;
+}
+
+static std::optional<std::string_view> LoadCompilerStubOnce(
+    const std::string& library_path) {
+  static std::string* loaded_path = [&]() -> std::string* {
+    if (InitializeCompilerForProcess(library_path)) {
+      return new std::string(library_path);
+    }
+    return nullptr;
+  }();
+
+  if (loaded_path) return *loaded_path;
+  return std::nullopt;
+}
+
+std::unique_ptr<OpenXlaCompiler> CreateOpenXlaCompiler() {
+  LoadCompilerStubOnce(GetIREECompilerPath());
+
+  auto* session = ireeCompilerSessionCreate();
+  auto* inv = ireeCompilerInvocationCreate(session);
+
+  ireeCompilerInvocationEnableConsoleDiagnostics(inv);
+
+  return std::make_unique<OpenXlaCompiler>(session, inv);
+}
```

Comment:
I'm wondering about the lifecycle here. Sessions can manage multiple invocations, so I could see an instance of `OpenXlaCompiler` persisting a session and creating invocations as needed.

(just authored this page recently)
https://openxla.github.io/iree/reference/bindings/c-api/#concepts

I'm not sure how much overhead is involved in creating and reusing a session... it might make sense to add some comments with rationale here. I could see the MLIRContext that a session contains persisting lots of memory (e.g. MLIR attributes) that would be useful to trim (such as by destroying the context between invocations, as done here).

Reply:
I copied this piece from openxla-pjrt-plugin without much understanding of what it does, I'll look into sharing a session in the followup changes